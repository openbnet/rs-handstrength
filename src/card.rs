#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub enum Suit {
    S,
    H,
    C,
    D,
    A, // any *
    E, // TrainerSuit first suit in cards
    F, //2nd suit
    G, // 3rd suit
    X, // Except E - 3 possible
    Y, // Except E and F - 2 possible
    Z, // Except E, F, and G -4th suit

}

#[derive(Clone, Copy, Debug, Eq, Hash,PartialOrd, PartialEq, Ord)]
pub struct Card {
    pub value: u8,
    pub suit: Suit,
}
use std::fmt;

impl fmt::Display for Suit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let suit = match self {
            Suit::C => "c",
            Suit::D => "d",
            Suit::H => "h",
            Suit::S => "s",
            Suit::A => "*",
            Suit::E => "e",
            Suit::F => "f",
            Suit::G => "g",
            Suit::X => "x",
            Suit::Y => "y",
            Suit::Z => "z",
        };
        write!(f, "{}", suit)
    }
}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} of {}", self.value, self.suit)
    }
}
impl Default for Card {
    fn default() -> Self {
        Card {
            value: 0,
            suit: Suit::S,
        }
    }
}