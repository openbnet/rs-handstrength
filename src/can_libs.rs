use crate::card::{Card, Suit};

pub type CardHand = Vec<Card>;
pub type SameRankHands = Vec<CardHand>;
pub type PEval = Vec<SameRankHands>;
pub type CanHaveCombis = Vec<(Vec<Card>, PEval)>;
pub fn can_flush(comcards: &Vec<Card>) -> Option<Suit> {
    let mut suit_counts = std::collections::HashMap::new();

    for card in comcards {
        *suit_counts.entry(card.suit).or_insert(0) += 1;
    }

    suit_counts.into_iter()
        .find(|&(_, count)| count >= 3)
        .map(|(suit, _)| suit)
}
pub fn can_flush_hand(hand: &Vec<Card>) -> Option<Suit> {
    let mut suit_counts = std::collections::HashMap::new();

    for card in hand {
        *suit_counts.entry(card.suit).or_insert(0) += 1;
    }

    suit_counts.into_iter()
        .find(|&(_, count)| count >= 2)
        .map(|(suit, _)| suit)
}

// cal gaps EXPECTS desc sorted combi of 3 only!
pub fn calculate_gaps(sorted_combination: &[Card]) -> (u8, Vec<Vec<u8>>) {
    let mut gaps: Vec<Vec<u8>> = Vec::new();
    let mut num_gaps = 0;

    for i in 0..sorted_combination.len() - 1 {
        let diff = sorted_combination[i].value as i32 - sorted_combination[i + 1].value as i32 - 1;
        if diff < 1 {
            continue;
        }

        let mut current_gap: Vec<u8> = Vec::new();
        for y in 1..=diff {
            current_gap.push((sorted_combination[i].value as i32 - y) as u8);
            num_gaps += 1;
        }

        gaps.push(current_gap);
    }

    (num_gaps as u8, gaps)
}
pub fn add_ace_as_high(cards: &Vec<Card>) -> Vec<Card> {
    let mut modified_cards = cards.clone();

    // For each Ace, add an additional Ace with value 14 and the same suit
    for card in cards.iter().filter(|&card| card.value == 1) {
        modified_cards.insert(0, Card { value: 14, suit: card.suit });
    }

    modified_cards
}
pub fn replace_ace_as_high(cards: &Vec<Card>) -> Vec<Card> {
    let mut modified_cards = Vec::new();

    // For each Ace, add an additional Ace with value 14 and the same suit
    for card in cards.iter() {
        if card.value != 1 {
            modified_cards.push(card.clone());
        } else {
            modified_cards.insert(0, Card { value: 14, suit: card.suit });
        }

    }
    modified_cards
}