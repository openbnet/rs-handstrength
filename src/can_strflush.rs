use crate::card::Card;
use crate::sorting::sort_cards;
use crate::can_libs::*;

pub fn can_have_straight_flush(comcards: &Vec<Card>) -> CanHaveCombis {
    let sorted = sort_cards(comcards);
    let mut straight_flush_combinations: CanHaveCombis = Vec::new();
    // println!("cah have str flush called {:?}", comcards);
    let flushsuit = can_flush( &sorted);
    // println!("flush suit {:?}", flushsuit);
    if let Some(flush_suit) = can_flush(&sorted) {
        let flush_suit_cards = add_ace_as_high(
            &sorted.iter()
                  .filter(|card| card.suit == flush_suit)
                  .cloned()
                  .collect()
        );
        let length = flush_suit_cards.len();
        for i in 0..length {
      
            for j in i + 1..length {
                for k in j + 1..length {
                    let combination = vec![flush_suit_cards[i], flush_suit_cards[j], flush_suit_cards[k]];
                    let gaps = calculate_gaps(&combination);
                    // println!("Combination: {:?}, Gaps: {:?}", combination, gaps);

                    if gaps.0 <= 2 {
                        match is_blocked(&combination, &flush_suit_cards, gaps) {
                            Ok(possible_combinations) => {
                                // println!("Not Blocked: {:?}", combination);
                                straight_flush_combinations.push((combination, possible_combinations));
                            },
                            Err(_) => {
                                // println!("Blocked: {:?}", combination);
                            }
                        }
                    } else {
                        // println!("Too many gaps: {:?}", combination);
                    }
                }
            }
        }
    }
    // println!("Straight Flush Combinations: {:?}", straight_flush_combinations);
    handle_ace_low(straight_flush_combinations)
}

pub fn handle_ace_low(combinations: CanHaveCombis) -> CanHaveCombis {
    let mut modified_combinations: CanHaveCombis = Vec::new();

    for (straight_flush_combination, blocking_combinations) in combinations {
        let mut modified_blocking_combinations = Vec::new();
        let contains_ace_high = straight_flush_combination.iter().any(|card| card.value == 14);
        let contains_ace_low = straight_flush_combination.iter().any(|card| card.value == 1);

        for combo in blocking_combinations {
            // Handle Ace as high (14) or low (1), but not both
            if contains_ace_high {
                // Convert Ace to high if it's part of the combination
                let ace_high_combo = convert_ace_to_high(&combo);
                modified_blocking_combinations.push(ace_high_combo);
            } else if contains_ace_low {
                // Convert Ace to low if it's part of the combination
                let ace_low_combo = convert_ace_to_low(&combo);
                modified_blocking_combinations.push(ace_low_combo);
            } else {
                // If no Ace is involved, keep the combination as is
                modified_blocking_combinations.push(combo);
            }
        }

        // Deduplicate the combinations
        modified_blocking_combinations.sort();
        modified_blocking_combinations.dedup();

        modified_combinations.push((straight_flush_combination, modified_blocking_combinations));
    }

    modified_combinations
}

// Helper functions to convert Ace to high or low
fn convert_ace_to_high(combo: &Vec<CardHand>) -> Vec<CardHand> {
    combo.iter().map(|hand| {
        hand.iter().map(|card| {
            if card.value == 1 { Card { value: 14, suit: card.suit } } else { *card }
        }).collect()
    }).collect()
}

fn convert_ace_to_low(combo: &Vec<CardHand>) -> Vec<CardHand> {
    combo.iter().map(|hand| {
        hand.iter().map(|card| {
            if card.value == 14 { Card { value: 1, suit: card.suit } } else { *card }
        }).collect()
    }).collect()
}

pub fn is_blocked(straight_flush_cards: &[Card],all_flush_cards: &Vec<Card>, gaps: (u8, Vec<Vec<u8>>)) -> Result<Vec<SameRankHands>, bool> {
    if gaps.0 > 2 {
        return Err(true); // Too many gaps, blocked by default
    }

    let values: Vec<u8> = straight_flush_cards.iter().map(|card| card.value).collect();
    let highest = *values.iter().max().unwrap();
    let lowest = *values.iter().min().unwrap();
    let suit = straight_flush_cards[0].suit; // Assuming all straight flush cards have the same suit
    let not_my_cards: Vec<Card> = all_flush_cards.iter()
        .filter(|&card| !straight_flush_cards.contains(card))
        .cloned()
        .collect();
    let not_my_values: Vec<u8> = not_my_cards.iter().map(|card| card.value).collect();
    // println!("is blocked start {:?} {:?}", values, not_my_cards);
    match gaps.0 {
        0 => {
            // Check for cards above and below the sequence
            let mut possible = Vec::new();
            // println!("0 gapper {:?} {:?}", values, not_my_values);
            if !not_my_values.contains(&(highest + 1)) && !not_my_values.contains(&(highest + 2)) {
                possible.push(vec![vec![
                    Card { value: highest + 1, suit },
                    Card { value: highest + 2, suit },
                ]]);
            }
            if lowest > 1 && highest < 14 && !not_my_values.contains(&(lowest - 1)) && !not_my_values.contains(&(highest +1)) {
                possible.push(vec![vec![
                    Card { value: highest + 1, suit },
                    Card { value: lowest - 1, suit },
                ]]); 
            }
            if lowest > 2 && !not_my_values.contains(&(lowest - 1)) && !not_my_values.contains(&(lowest - 2)) {
                possible.push(vec![vec![
                    Card { value: lowest - 1, suit },
                    Card { value: lowest - 2, suit },
                ]]);
            }

            if possible.is_empty() { Err(true) } else { Ok(possible) }
        },
        1 => {
            // Handle the single gap
            let gap_value = gaps.1[0][0];
            if not_my_values.contains(&gap_value) {
                Err(true)
            } else {
                let mut possible = Vec::new();
                if highest < 14 && !not_my_values.contains(&(highest + 1)) {
                    possible.push(vec![vec![
                        Card { value: highest + 1, suit },
                        Card { value: gap_value, suit },
                    ]]);
                }
                if lowest > 1 && !not_my_values.contains(&(lowest - 1)) {
                    possible.push(vec![vec![
                        Card { value: gap_value, suit },
                        Card { value: lowest - 1, suit },
                    ]]);
                }
                if possible.is_empty() { Err(true) } else { Ok(possible) }
            }
        },
        2 => {
            // Handle two 1-gappers or one 2-gapper
            let first_gap = &gaps.1[0];

            let mut possible = Vec::new();
            // println!("is blocked 2 {:?} {:?} {:?} first gap {:?}", gaps, gaps.1, gaps.1.len(), first_gap[0]);
            // Check if it's two 1-gappers or one 2-gapper
            if first_gap.len() == 2 {
                // It's a single 2-gapper
                possible.push(Card { value: first_gap[0], suit });
                possible.push(Card { value: first_gap[1], suit });
            } else {
                // It's two separate 1-gappers
                let second_gap = &gaps.1[1];
                // println!("second_gap {:?}", second_gap[0]);
                possible.push(Card { value: first_gap[0], suit });
                possible.push(Card { value: second_gap[0], suit });
            }

            if possible.is_empty() { Err(true) } else { Ok(
                vec![vec![possible]]
            ) }
        },
        _ => Err(true) // Any other case is considered blocked
    }
}