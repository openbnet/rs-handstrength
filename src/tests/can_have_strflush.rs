#[cfg(test)]
mod tests {
    use crate::card::{Card, Suit};
    use crate::can_strflush::*;
    use crate::can_libs::*;
    use insta::assert_debug_snapshot;
    #[test]
    fn no_strflush() {
        let comcards = vec![
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::H },
            Card { value: 4, suit: Suit::D },
            Card { value: 5, suit: Suit::C }
        ];
        assert!(can_have_straight_flush(&comcards).is_empty());
    }
    #[test]
    fn one_strflush() {
        let comcards = vec![
            Card { value: 3, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 4, suit: Suit::S },
            Card {value: 1, suit: Suit::S }
        ]; 
        let strflush = can_have_straight_flush(&comcards);
        println!("strflush {:?}", strflush);
        assert_eq!(strflush.len(), 1);
        let can_have = can_have_straight_flush(&comcards);
        println!("can_have {:?}", can_have);
        assert_debug_snapshot!(can_have); 
    }
    #[test]
    fn cal_gaps1() {
        let comcards = [
            Card { value: 4, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 1, suit: Suit::S },

        ]; 
        let v = vec![
            Card { value: 4, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 1, suit: Suit::S }, 
        ];

        let gaps = calculate_gaps(&comcards);
        // println!("gaps, {:?}", gaps1);
        assert_eq!(gaps.0, 1);
        assert_eq!(gaps.1, vec![vec![3]]);
        let blocked = is_blocked(&comcards, &v, gaps);
        println!("blocked 1 {:?}", blocked);
        assert_debug_snapshot!(blocked);
        let can_have = can_have_straight_flush(&v);
        println!("can_have {:?}", can_have);
        assert_debug_snapshot!(can_have); 
    }
    #[test]
    fn strflush1() {
        let comcards = vec![
            Card { value: 4, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 1, suit: Suit::S },
        ]; 
        let can_have = can_have_straight_flush(&comcards);
        println!("can_have {:?}", can_have);
        assert_debug_snapshot!(can_have);
        
    }
    #[test]
    fn strflush2() {
        let comcards = vec![
            Card { value: 4, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::S },
        ]; 
        let can_have = can_have_straight_flush(&comcards);
        println!("can_have {:?}", can_have);
        assert_debug_snapshot!(can_have);
    }
    #[test]
    fn cal_gaps21() {
        let comcards = [
            Card { value: 5, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 1, suit: Suit::S },

        ]; 
        let v = vec![
            Card { value: 5, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 1, suit: Suit::S }, 
        ];
        let gaps = calculate_gaps(&comcards);
        println!("gaps, {:?}", gaps);
        assert_eq!(gaps.0, 2);
        assert_eq!(gaps.1, vec![vec![4, 3]]);
        let blocked = is_blocked(&comcards, &v, gaps);
        println!("blocked 21 {:?}", blocked);
        assert_debug_snapshot!(blocked);
        let can_have = can_have_straight_flush(&v);
        println!("can_have {:?}", can_have);
        assert_debug_snapshot!(can_have);
    }
    #[test]
    fn cal_gaps12() {
        let comcards = [
            Card { value: 5, suit: Suit::S },
            Card { value: 3, suit: Suit::S },
            Card { value: 1, suit: Suit::S },

        ]; 
        let v = vec![
            Card { value: 5, suit: Suit::S },
            Card { value: 3, suit: Suit::S },
            Card { value: 1, suit: Suit::S },

        ]; 
        let gaps = calculate_gaps(&comcards);
        println!("gaps, {:?}", gaps);
        assert_eq!(gaps.0, 2);
        assert_eq!(gaps.1, vec![vec![4], vec![2]]);
        let blocked = is_blocked(&comcards, &v, gaps);
        println!("blocked {:?}", blocked);
        assert_debug_snapshot!(blocked);
        let can_have = can_have_straight_flush(&v);
        println!("can_have {:?}", can_have);
        assert_debug_snapshot!(can_have); 
    }
    #[test]
    fn cal_gaps31() {
        let comcards = [
            Card { value: 9, suit: Suit::S },
            Card { value: 5, suit: Suit::S },
            Card { value: 4, suit: Suit::S },
        ]; 
        let v = vec![
            Card { value: 9, suit: Suit::S },
            Card { value: 5, suit: Suit::S },
            Card { value: 4, suit: Suit::S },
        ];
        let gaps = calculate_gaps(&comcards);
        println!("gaps, {:?}", gaps);
        assert_eq!(gaps.0, 3);
        assert_eq!(gaps.1, vec![vec![8, 7, 6]]);
        let can_have = can_have_straight_flush(&v);
        println!("can_have {:?}", can_have);
        assert_eq!(can_have.len(), 0);
    }
}