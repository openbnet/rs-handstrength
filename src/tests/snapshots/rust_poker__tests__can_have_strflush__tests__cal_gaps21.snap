---
source: src/tests/can_have_strflush.rs
expression: blocked
---
Ok(
    [
        [
            Card {
                value: 4,
                suit: S,
            },
            Card {
                value: 3,
                suit: S,
            },
        ],
    ],
)
